adapter-base/src/main/java/org/eclipse/hono/adapter/MultiMessagingSenders.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2021 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.eclipse.hono.adapter;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import org.eclipse.hono.util.TenantConstants;
+import org.eclipse.hono.util.TenantObject;
+
+import io.vertx.core.json.JsonObject;
+
+/**
+ * Encapsulates the Kafka-based and the AMQP-based sender for a message type.
+ *
+ * @param <T> The sender type (intended to be one of {@link org.eclipse.hono.adapter.client.telemetry.EventSender},
+ *            {@link org.eclipse.hono.adapter.client.telemetry.TelemetrySender}, or
+ *            {@link org.eclipse.hono.adapter.client.command.CommandResponseSender}).
+ */
+class MultiMessagingSenders<T> {
+
+    private T amqpSender;
+    private T kafkaSender;
+
+    /**
+     * Sets the AMQP-based sender.
+     *
+     * @param amqpSender The sender.
+     * @throws NullPointerException if amqpSender is {@code null}.
+     */
+    public void setAmqpSender(final T amqpSender) {
+        this.amqpSender = Objects.requireNonNull(amqpSender);
+    }
+
+    /**
+     * Sets the Kafka-based sender.
+     *
+     * @param kafkaSender The sender.
+     * @throws NullPointerException if kafkaSender is {@code null}.
+     */
+    public void setKafkaSender(final T kafkaSender) {
+        this.kafkaSender = Objects.requireNonNull(kafkaSender);
+    }
+
+    /**
+     * Gets the AMQP-based sender.
+     *
+     * @return The sender or {@code null}.
+     */
+    public T getAmqpSender() {
+        return amqpSender;
+    }
+
+    /**
+     * Gets the Kafka-based sender.
+     *
+     * @return The sender or {@code null}.
+     */
+    public T getKafkaSender() {
+        return kafkaSender;
+    }
+
+    /**
+     * Checks if none of the senders are set.
+     *
+     * @return true if both senders are {code null}.
+     */
+    public boolean isUnconfigured() {
+        return amqpSender == null && kafkaSender == null;
+    }
+
+    /**
+     * Gets the client to be used for sending messages downstream.
+     *
+     * @param tenant The tenant for which to send messages.
```

Comment:
FMPOV we also need to document here which properties of the tenant will determine the returned type of sender